server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/rust/RustLanguagePlugin.kt

```diff
@@ -0,0 +1,331 @@
+                location = location,
+                fromWorkspace = targetInfo.rustCrateInfo.fromWorkspace,
+                name = targetInfo.rustCrateInfo.name,
+                kind = targetInfo.rustCrateInfo.kind,
+                edition = targetInfo.rustCrateInfo.edition,
+                crateFeatures = targetInfo.rustCrateInfo.crateFeaturesList,
+                dependencies = targetInfo.rustCrateInfo.dependenciesList.mapNotNull {
+                    RustDependency(
+                        crateId = it.crateId,
+                        rename = it.rename,
+                    )
+                },
+                crateRoot = crateRoot(),
+                version = targetInfo.rustCrateInfo.version,
+                procMacroArtifacts = targetInfo.rustCrateInfo.procMacroArtifactsList,
+                procMacroSrv = targetInfo.rustCrateInfo.procMacroSrv,
+                rustcSysroot = targetInfo.rustCrateInfo.rustcSysroot,
+                rustcSrcSysroot = targetInfo.rustCrateInfo.rustcSrcSysroot,
+                cargoBinPath = targetInfo.rustCrateInfo.cargoBinPath,
+                rustcVersion = targetInfo.rustCrateInfo.rustcVersion,
+            )
+        }
+    }
+
+    private fun TargetInfo.getRustCrateInfoOrNull(): BspTargetInfo.RustCrateInfo? =
+        this.takeIf(TargetInfo::hasRustCrateInfo)?.rustCrateInfo
+
+    override fun applyModuleData(moduleData: RustModule, buildTarget: BuildTarget) {
+        // TODO
+    }
+
+    data class BazelPackageTargetInfo(val packageName: String, val targetName: String)
+
+    private fun resolvePackage(rustTarget: Module): BazelPackageTargetInfo {
+        return resolvePackage(rustTarget.label)
+    }
+
+    private fun resolvePackage(label: Label): BazelPackageTargetInfo {
+        val labelVal = label.value
+        val packageName = labelVal.substringBefore(":")
+        val targetName = labelVal.substringAfter(":")
+        return BazelPackageTargetInfo(packageName, targetName)
+    }
+
+    // We need to merge targets with the same crate root.
+    private fun mergeTargets(rustTargets: List<Pair<Module, RustModule>>): Pair<Module, RustModule> {
+        if (rustTargets.size == 1) {
+            return rustTargets[0]
+        }
+
+        // TODO:
+        // We need some heuristic here. As we need the proper label, lets take the target with
+        //  the most features. It can be changes as a union of features set, but we need
+        //  to give it a proper label.
+
+        return rustTargets.maxByOrNull { (_, rustData) -> rustData.crateFeatures.size }!!
+    }
+
+    private fun prependWorkspacePath(path: String): String =
+        bazelPathsResolver.relativePathToExecRootAbsolute(path).toString()
+
+    private fun rustPackages(rustBspTargets: List<Module>): List<RustPackage> {
+        for (target in rustBspTargets) {
+            require(Language.RUST in target.languages) { "The target is not a Rust target" }
+        }
```

Comment:
prob we dont need it - it's a private methods and should know on what data we call ut

Reply:
done